Adaptations for macOS 11.0

csomor opened the pull request:
This PR should be used for all API extensions or code adaptions needed for Apple Silicon

csomor committed:
Stop using obsolete AbsoluteToNanoseconds()

csomor committed:
Stop using deprecated CFAbsoluteTimeAddGregorianUnits()

since we are still working in seconds, the replacement is rather simple

csomor committed:
Removing unused obsolete code from wxFileDialog

We don't need workaround for macOS earlier than 10.6 any longer.

vadz commented:
No real comments, thanks for clearning this up, but it would be nice to have a bit more details in the commit messages as it's not really clear to me why is `wxOpenPanelDelegate` being removed -- but then it looks like it wasn't used anyhow? If this is the case, I'd mention this in the commit message to preempt any questions.

csomor commented:
Sorry, yes the obsolete code was not used anymore

csomor committed:
Show using window-modal file open dialog in the sample

csomor committed:
Show all other supported window modal dialogs in the sample too

csomor committed:
Replace deprecated API usage for window modal dialogs

Remove the delegate objects that are not used any longer.

Move wxGetAvailableDrives() to Objective-C code file dirdlg.mm to allow
for Cocoa implementation.

imciner2 commented:
@csomor many of your changes to the `dirdlg.mm` file are already done in https://github.com/wxWidgets/wxWidgets/pull/1884 which is just waiting for @vadz to merge.

csomor commented:
> @csomor many of your changes to the `dirdlg.mm` file are already done in #1884 which is just waiting for @vadz to merge.

Thanks, I'll compare

csomor committed:
Removing unused file-related functions

This should be done back in 5ba67c67e4 (Remove wxOSX/Carbon support.,
2016-01-22).

csomor committed:
Replace deprecated LSGetApplicationForInfo()

vadz commented:
> > @csomor many of your changes to the `dirdlg.mm` file are already done in #1884 which is just waiting for @vadz to merge.
> 
> Thanks, I'll compare

Just FYI: I've extracted this particular change in a separate commit b98660b996bfdf39784259228b5c7d0ff3cad646

csomor commented:
> Just FYI: I've extracted this particular change in a separate commit [b98660b](https://github.com/wxWidgets/wxWidgets/commit/b98660b996bfdf39784259228b5c7d0ff3cad646)

perfect, thanks, I'll extract the things that are better in his version

csomor committed:
Avoid infinite recursion between wxStockGDI and wxSystemSettions

There was a possible recursion from wxStockGDI::GetFont to
wxSystemSettings::GetFont and back, resolve by using direct font
creation, adding fixed system font.

csomor committed:
Stop using deprecated API for assert output handler

csomor committed:
Modernize accessibility API usage

csomor committed:
Replace deprecated NSDate dateWithString

csomor committed:
Produce NFD in wxConvFileName::FromWChar() on macOS

This ensures that fn_str() returns the string in the expected,
decomposed, format.

Also simplify the code by removing workarounds for old systems which are
not supported any more and make explicit the fact that under macOS
ToWChar() always produced NFC.

vadz commented on include/wx/osx/core/private/strconv_cf.h:
```diff
@@ -316,9 +316,10 @@ class wxMBConv_cf : public wxMBConv
     }
 
-    void Init( CFStringEncoding encoding)
+    void Init( CFStringEncoding encoding, bool compose_decompose = false)
```
This is nitpicking, of course, but I'd strongly prefer having
```
enum NormalForm
{
    NF_Composed,
    NF_Decomposed
};
```
instead because `Init(..., true)` is really unreadable.

csomor replied:
I didn't want this to be in the public API, and it would have been for both composing and decomposing, so it would be a switch on and off, composing when reading in from utf8, decomposing when writing out to utf8. Then  I'd like to make this a bit field, having one bit setting composing when reading in and another when writing out, that way we can always decide

csomor replied:
So the whole range would become something like this

```
    enum NormalizationForm {
        None = 0x00,
        
        FromWChar_C = 0x02,
        FromWChar_D = 0x03,
        FromWChar_Mask = 0x03,
        
        ToWChar_C = 0x08,
        ToWChar_D = 0x0C,
        ToWChar_Mask = 0x0C
    };
```

so the default to preserve behavior would be `ToWChar_C`, for wxConvFilename this would be `ToWChar_C |  FromWChar_D`,

what do you think about that ? and do we have a template somewhere for bitwise operators on enums ?

vadz replied:
I don't mind but, to be honest, this looks like overkill to me as I don't think we actually need all the possible combinations, I'd just keep using NFC internally because this is the least surprising form and provide an option to use NFD for macOS file (and maybe others, I still don't know if it's also needed elsewhere) functions.

We don't have any template for enum bit operations and IMO it wouldn't be worth it to write it just for this enum.

csomor replied:
I'm ok with keeping the behavior NFC by default, but I'd like to make it explicit in the API, so I'll just keep 

    enum NormalizationForm {
        None = 0x00,
        FromWChar_C = 0x01,
        ToWChar_D = 0x02,
    };

vadz commented on include/wx/osx/core/private/strconv_cf.h:
```diff
@@ -333,9 +334,23 @@ class wxMBConv_cf : public wxMBConv
+    {
+        Init( wxCFStringEncFromFontEnc(encoding) , true );
+    }
+
+    virtual ~wxMBConvD_cf()
+    {
+    }
```
```suggestion
```

vadz commented on include/wx/osx/core/private/strconv_cf.h:
```diff
@@ -333,9 +334,23 @@ class wxMBConv_cf : public wxMBConv
+class wxMBConvD_cf : public wxMBConv_cf
+{
+public:
+    wxMBConvD_cf(wxFontEncoding encoding)
```
```suggestion
    explicit wxMBConvD_cf(wxFontEncoding encoding)
```

vadz commented on src/osx/core/strconv_cf.cpp:
```diff
@@ -89,27 +89,23 @@ WXDLLIMPEXP_BASE wxMBConv* new_wxMBConv_cf(wxFontEncoding encoding)
         if ( theString == NULL )
             return wxCONV_FAILED;
 
+        if ( m_compose_decompose )
```
I don't really understand why we don't do this for "plain" `wxMBConv_cf` any more... I think the comment below is still relevant, i.e. we do want to use NFC internally -- but now we don't, unless we're working with file paths.

csomor replied:
Wasn't the reason for normalizing the input from filenames, so this would be wxConvFilename ? I thought that having round-trip fidelity in the normal cf conversion would be good. But I can leave it on of course

vadz replied:
I'm not sure why was this done initially, but, as I wrote on the list, `wxMBConv_cf` is used by `wxCSConv`, i.e. this is going to change the behaviour of any code using this (public) class and it doesn't look like a good idea, especially because the old behaviour (outputting NFC) seems more useful to me.

For round-trip fidelity, we could (should?) consider using NFC in `ToWChar()` too, unless the new flag (bool or enum) indicates that NFD should be used.

csomor replied:
I was thinking about roundtrip fidelity (not for the filename converter, but the 'normal' one) in the sense that a decomposed utf8 would be still decomposed in the wxString, and then when written back it would still be decomposed, exactly as a composed utf8 would be composed wxString and written out composed. So maintaining the normalization throughout the conversion.

APFS once had a version that that only opened the files if the normalization was correct, so as with case sensitive filesystem, it was normalization sensitive. You could have to identical looking filenames one composed one decomposed, eg one typed in the finder decomposed, another one copied from a windows file system composed. You had to preserve normalization otherwise you would open the wrong file. But this situation is better now, it is now normalization-insensitive ...

vadz replied:
Note that `wxString` never stores UTF-8 in the default build (I agree that what you say would be a concern for `wxUSE_UNICODE_UTF8` build, but AFAIK nobody uses it anyhow, so...). This is why I say that NFC only/mostly matters in `ToWChar()`.

vadz commented on src/osx/core/strconv_cf.cpp:
```diff
@@ -89,27 +89,23 @@ WXDLLIMPEXP_BASE wxMBConv* new_wxMBConv_cf(wxFontEncoding encoding)
          * in the count iff it was included in the source count.
          */
 
-
```
These simplifications are definitely very nice, it just would have been great to have them in a separate commit to make things more clear...

vadz reviewed (commented):
Sorry if I'm missing something but removing `CFStringNormalize(kCFStringNormalizationFormC)` call from the normal code path really doesn't seem right to me.

And, of course, I [hate bool params](https://www.wxwidgets.org/develop/coding-guidelines/#no_bool_params).

csomor commented:
@vadz what do you think, is it ok when I squash and merge this PR ?

vadz commented:
I think there are several commits here that stand well on their own and it would definitely make reviewing changes in the future easier if we kept them. Again, if you'd like, I could squash/rebase things here (without changing anything except maybe formatting) and merge it, please let me know if you'd like me to do it.

csomor commented:
this would be very helpful, thanks, as I'm not really fluent yet in merging things in separate batches ...

vadz commented:
Stefan, I'm going through this and [debug macros commit](https://github.com/wxWidgets/wxWidgets/pull/1940/commits/48bc7fde072f2183dcb2e5038582ea62f265d498) just doesn't make sense to me: `wxMacAssertOutputHandler()` doesn't seem to be used anywhere and neither does `DEBUG_ASSERT_MESSAGE` and I don't think we should define `DEBUG_ASSERT_PRODUCTION_CODE` in `wx/debug.h` anyhow, as it should be application-level decision.

Could we perhaps just omit this for now, i.e. simply remove `InstallDebugAssertOutputHandler()` call which is, AFAICS, was the trigger for these changes?

vadz committed:
Remove the now unnecessary empty wxMessageDialog dtor

No real changes, just some minor cleanup after 220dfe17ea (Replace
deprecated API usage for window modal dialogs, 2020-07-09).

vadz commented:
I went ahead and force-pushed my branch. I didn't make (or at least tried not to make) any changes here, all I did was to squash together commits that corrected mistakes in earlier ones and changed the commit messages to be in the standard imperative mood and capitalized them for consistency.

I did remove most of the debug macros commit as I'm just not sure if it's really correct to do this and would like to discuss this a bit more.

And I've added a trivial last commit.

csomor commented:
> I did remove most of the debug macros commit as I'm just not sure if it's really correct to do this and would like to discuss this a bit more.

yes, it was connected to the removal of `InstallDebugAssertOutputHandler`, the defines were defined in a way that the native Asserts like eg `__Verify_noErr`  are still routed through our own wx assert messages when in debug builds and not just output to the console, because that way we can debug things

vadz commented:
I think that we should define `DEBUG_ASSERT_MESSAGE` in `wx/osx/private/debug.h` or something like this, i.e. I don't think we should impose it on the user code like this.

In fact, I don't really know why should we use Apple macros in our own code at all, instead of the standard `wxCHECK_XXX()`. Looking at the sources, I only see `__Verify_noErr` in `src/osx/carbon/region.cpp` and `src/osx/core/bitmap.cpp`, could we just ger rid of them?

csomor commented:
It has to be defined before any system headers are included, therefore I moved it there, but I'll check on completely replacing them. I'd like to keep the feature of these asserts - translating OS error codes to Text and going through or common assert method

vadz commented:
> It has to be defined before any system headers are included, therefore I moved it there, 

I am not sure if this worked neither, we include `wx/platform.h` before `wx/debug.h` from `wx/defs.h` and the it includes some macOS headers.

> but I'll check on completely replacing them. I'd like to keep the feature of these asserts - translating OS error codes to Text and going through or common assert method

Surely there are functions or macros for doing this too that we could use in our own macros?

Anyhow, I just don't think it's right for a library, even if it's a framework, to take control of such global symbols, they should really be left for the application IMNSHO.

vadz commented:
Should we merge this already (as the CI builds have finally passed) and deal with the debug macros in a separate PR later?

csomor commented:
yes, please, I'll let you push the right merge buttons ... otherwise I'll manage to sqash things ...

csomor commented:
Ok, the debug macros evolution will be in this PR

https://github.com/wxWidgets/wxWidgets/pull/1973
